Linkify URLs and issues in code comments

paulmolluzzo opened the pull request:
Closes #380 

This linkifies numbered issues (i.e. `#123`) and URLs found in code comments. You can take a look at this page to see it in action: https://github.com/contao/core/blob/549738f9a41974d963bd09a3a83ae0f46f887f01/system/modules/core/pages/PageForward.php#L51

In that page there is a URL and there are a few issues linked.

Should work on any page with code, so a PR with inline comments works: https://github.com/sindresorhus/refined-github/pull/346

The files on a PR: https://github.com/sindresorhus/refined-github/pull/346/files

And the code for individual files/commits: https://github.com/sindresorhus/refined-github/commit/2d0c354b66556cd15ac9f1fc5f775a313f026f96#diff-0730bb7c2e8f9ea2438b52e419dd86c9

// @Loilo

sindresorhus commented:
This is awesome!!

Another example: https://github.com/avajs/ava/blob/2b9f29133fc8306786615cae90ff0ebea0e3de28/test/runner.js#L191

sindresorhus commented:
Seems like it doesn't just linkify in comments, but anywhere. For example: https://github.com/sindresorhus/refined-github/commit/2d0c354b66556cd15ac9f1fc5f775a313f026f96#diff-0730bb7c2e8f9ea2438b52e419dd86c9 I think it makes sense to linkify URLs anywhere, but issues should only be linkified in comments as they can have many false positives.

sindresorhus commented:
@paulmolluzzo Can you add it to the feature highlight in the readme?

sindresorhus commented:
Could you also support the issue reference format `sindresorhus/refined-github#381`. It's useful for referencing other repos. See: https://guides.github.com/features/mastering-markdown/

sindresorhus commented:
Should the link maybe open in a new window/tab (`_blank`)?

paulmolluzzo commented:
> Could you also support the issue reference format sindresorhus/refined-github#381. It's useful for referencing other repos. See: https://guides.github.com/features/mastering-markdown/

@sindresorhus Do you have an example link with this in it?

Loilo commented:
I'd vote against the `_blank` option. I think folks using the extension are advanced enough to have a quick&easy way to open links in new tabs anyway while `_blank` would just take the option for the same tab away.

sindresorhus commented:
> Do you have an example link with this in it?

https://github.com/sindresorhus/playground/commit/faf058ff8af7283d29ecb8f3314ac4b3f5c8c2b2

paulmolluzzo commented:
I think I addressed everything. Please take a look. (I used all the URLs in this issue.)

I added the `_blank` request, but can remove if it's annoying.

paulmolluzzo commented:
Just pushed another small fix

paulmolluzzo commented:
> @paulmolluzzo Can you add it to the feature highlight in the readme?

@sindresorhus do you like any of these images for the feature in the README? If not and you have a better one in mind, please share and I'll update the README. For now it links to issue #380

![screen shot 2017-04-24 at 5 05 03 pm](https://cloud.githubusercontent.com/assets/737065/25358803/b3d8a25e-2910-11e7-9dcc-940831388bf4.png)
![screen shot 2017-04-24 at 5 08 23 pm](https://cloud.githubusercontent.com/assets/737065/25358802/b3d46fae-2910-11e7-8bc9-b3068a26956e.png)
![screen shot 2017-04-24 at 5 08 29 pm](https://cloud.githubusercontent.com/assets/737065/25358805/b3da1e18-2910-11e7-8152-4ed09b54b874.png)
![screen shot 2017-04-24 at 5 08 39 pm](https://cloud.githubusercontent.com/assets/737065/25358804/b3d8d7a6-2910-11e7-9dd9-a35965cb906e.png)

paulmolluzzo commented:
The tests make a good screenshot.

![screen shot 2017-04-24 at 10 03 51 pm](https://cloud.githubusercontent.com/assets/737065/25365902/f407f964-2939-11e7-9c66-7d48f31c5576.png)

sindresorhus commented on readme.md:
```diff
@@ -41,6 +41,7 @@ Our hope is that GitHub will notice and implement some of these much needed impr
 - Removes tooltips
 - Copy canonical link to file when [the `y` hotkey](https://help.github.com/articles/getting-permanent-links-to-files/) is used
 - ~~[Adds blame links for parent commits in blame view](https://github.com/sindresorhus/refined-github/issues/2#issuecomment-189141373)~~ [Implemented by GitHub](https://github.com/blog/2304-navigate-file-history-faster-with-improved-blame-view)
+- [Linkifies URLs and issue references in code](https://cloud.githubusercontent.com/assets/737065/25365902/f407f964-2939-11e7-9c66-7d48f31c5576.png)
```
Add it much closer to the top. Around line 20. This feature is awesome.

Maybe we should split it into two feature highlights?

1. Linkifies issue references in code comments
2. Linkifies URLs in code and comments

^ Or something better.

sindresorhus commented:
How about this:

<img width="899" alt="screen shot 2017-04-25 at 12 29 03 annotated" src="https://cloud.githubusercontent.com/assets/170270/25370217/61718820-29b3-11e7-89c5-2959eaf8cac8.png">

paulmolluzzo commented on readme.md:
```diff
@@ -14,6 +14,8 @@ Our hope is that GitHub will notice and implement some of these much needed impr
 - [Linkifies branch references in pull requests](https://github.com/sindresorhus/refined-github/issues/1)
 - [Linkifies issue/PR references in issue/PR titles](https://cloud.githubusercontent.com/assets/170270/13597190/bd487ec4-e549-11e5-9521-419fa284512c.png)
+- [Linkifies URLs in code](https://cloud.githubusercontent.com/assets/170270/25370217/61718820-29b3-11e7-89c5-2959eaf8cac8.png)
+- [Linkifies issue references in code comments](https://cloud.githubusercontent.com/assets/170270/25370217/61718820-29b3-11e7-89c5-2959eaf8cac8.png)
```
@sindresorhus I put it right under the other "linkifying" features and split it up as you suggested.

paulmolluzzo commented:
Fixed an issue with links getting duplicated.

sindresorhus commented:
Awesome work @paulmolluzzo. I've been using it for a day now and already found it useful many times! 🙌🎉

paulmolluzzo commented:
This is a great idea! Thanks @Loilo!
